File: src/ios/GoogleMaps/PluginDirectionsService.m
//
//  PluginElevationService.m
//  cordova-googlemaps-plugin v2
//
//  Created by Masashi Katsumata.
//
//

#import "PluginDirectionsService.h"

@implementation PluginDirectionsService

- (void)pluginUnload
{
  if (self.executeQueue != nil){
      self.executeQueue.suspended = YES;
      [self.executeQueue cancelAllOperations];
      self.executeQueue.suspended = NO;
      self.executeQueue = nil;
  }
}

- (void)pluginInitialize
{
  self.executeQueue = [NSOperationQueue new];
  self.executeQueue.maxConcurrentOperationCount = 3;

}

- (NSString *)_decode_DirectionsRequestLocation: (NSDictionary *)position {
  NSString *positionType = [position objectForKey:@"type"];
  
  if ([@"string" isEqualToString: positionType]) {
    return [[position objectForKey:@"value"] stringValue];
  }
  
  NSDictionary *value = [position objectForKey:@"value"];
  if ([@"location" isEqualToString:positionType]) {
    return [NSString stringWithFormat:@"%f,%f",
              [[value objectForKey:@"lat"] doubleValue],
              [[value objectForKey:@"lng"] doubleValue]
            ];
  }
  if ([value objectForKey:@"placeId"]) {
    return [NSString stringWithFormat:@"place_id:%@",
              [[value objectForKey:@"placeId"] stringValue]
            ];
  }
  
  if ([value objectForKey:@"location"]) {
    NSDictionary *location = [value objectForKey:@"location"];
    return [NSString stringWithFormat:@"%f,%f",
              [[location objectForKey:@"lat"] doubleValue],
              [[location objectForKey:@"lng"] doubleValue]
            ];
  }
  if ([value objectForKey:@"query"]) {
    return [value objectForKey:@"query"];
  }
  return @"";
}

- (void)route:(CDVInvokedUrlCommand*)command {
  NSDictionary *opts = [command.arguments objectAtIndex:0];
  NSLog(@"route = %@", opts);
  

  NSMutableDictionary *params = [NSMutableDictionary dictionary];
  //-----------------------
  // required parameters
  //-----------------------
  [params setObject:[self _decode_DirectionsRequestLocation:[opts objectForKey:@"origin"]] forKey:@"origin"];
  [params setObject:[self _decode_DirectionsRequestLocation:[opts objectForKey:@"destination"]] forKey:@"destination"];
  
  //-----------------------
  // mode parameter
  //-----------------------
  if ([opts objectForKey:@"travelMode"]) {
    // Default: driving
    NSString *travelMode = [opts objectForKey:@"travelMode"];
    travelMode = [travelMode lowercaseString];
    if (![@"driving" isEqualToString:travelMode]) {
      [params setObject:travelMode forKey:@"mode"];
      
      //-----------------------
      // transitOptions parameter
      //-----------------------
      if ([@"transit" isEqualToString:travelMode] &&
          [opts objectForKey:@"transitOptions"]) {
        NSDictionary *transitOptions = [opts objectForKey:@"transitOptions"];
        if ([transitOptions objectForKey:@"arrivalTime"]) {
          [params setObject:[NSString stringWithFormat:@"%@", [transitOptions objectForKey:@"arrivalTime"]] forKey:@"arrival_time"];
        }
        if ([transitOptions objectForKey:@"depatureTime"]) {
          [params setObject:[NSString stringWithFormat:@"%@", [transitOptions objectForKey:@"depatureTime"]] forKey:@"depature_time"];
        }
      }
    } else {
       //-----------------------
       // DrivingOptions parameter
       //-----------------------
       if ([opts objectForKey:@"drivingOptions"]) {
         NSDictionary *drivingOptions = [opts objectForKey:@"drivingOptions"];
         if ([drivingOptions objectForKey:@"depatureTime"]) {
           [params setObject:[NSString stringWithFormat:@"%@", [drivingOptions objectForKey:@"depatureTime"]] forKey:@"depature_time"];
         }
         if ([drivingOptions objectForKey:@"trafficModel"]) {
           NSString *trafficModel = [[[opts objectForKey:@"trafficModel"] stringValue] lowercaseString];
           [params setObject:trafficModel forKey:@"traffic_model"];
         }
       }
    }

    //-------------------
    // transitOptions parameter
    //-------------------
    if ([opts objectForKey:@"transitOptions"]) {
      NSDictionary *transitOptions = [opts objectForKey:@"transitOptions"];
      //-------------------
      // transitOptions.modes parameter
      //-------------------
      if ([transitOptions objectForKey:@"modes"]) {
        NSArray *modes = [transitOptions objectForKey:@"modes"];
        NSString *buffer = @"";
        for (int i = 0; i < modes.count; i++) {
          if (i > 0) {
            buffer = [buffer stringByAppendingString:@"|"];
          }
          buffer = [buffer stringByAppendingString:[[modes objectAtIndex:i] lowercaseString]];
        }
        [params setObject:buffer forKey:@"transit_mode"];
      }
      //-------------------
      // transitOptions.routingPreference parameter
      //-------------------
      if ([transitOptions objectForKey:@"routingPreference"]) {
        NSString *routingPreference = [[[transitOptions objectForKey:@"routingPreference"] stringValue] lowercaseString];
        [params setObject:routingPreference forKey:@"transit_routing_preference"];
      }
    }
  }
  //-------------------
  // waypoints parameter
  //-------------------
  if ([opts objectForKey:@"waypoints"]) {
    NSArray *waypoints = [opts objectForKey:@"waypoints"];
    NSString *buffer = @"";
    int cnt = 0;
    if ([opts objectForKey:@"optimizeWayPoints"]) {
      cnt = 1;
      buffer = @"optimize:true";
    }
    for (int i = 0; i < waypoints.count; i++) {
      NSDictionary *point = [waypoints objectAtIndex:i];
      bool stopOver = false;
      if ([point objectForKey:@"location"]) {
        if ([point objectForKey:@"stopover"]) {
          stopOver = [point objectForKey:@"stopover"];
        }
        if (cnt > 0) {
          buffer = [buffer stringByAppendingString:@"%7C"];
        }
        if (!stopOver) {
          buffer = [buffer stringByAppendingString:@"via:"];
        }
        buffer = [buffer stringByAppendingString:[self _decode_DirectionsRequestLocation:[point objectForKey:@"location"]]];
        cnt++;
      }
    }
    if (cnt > 0) {
      [params setObject:buffer forKey:@"waypoints"];
    }
  }
  //-------------------
  // alternatives parameter
  //-------------------
  if ([opts objectForKey:@"provideRouteAlternatives"]) {
    [params setObject:@"true" forKey:@"alternatives"];
  }
  
  //-------------------
  // avoid parameter
  //-------------------
  bool avoidFerries = false;
  bool avoidHighways = false;
  bool avoidTolls = false;
  if ([opts objectForKey:@"avoideFerries"]) {
    avoidFerries = [[opts objectForKey:@"avoideFerries"] boolValue];
  }
  if ([opts objectForKey:@"avoidHighways"]) {
    avoidHighways = [[opts objectForKey:@"avoidHighways"] boolValue];
  }
  if ([opts objectForKey:@"avoidTolls"]) {
    avoidTolls = [[opts objectForKey:@"avoidTolls"] boolValue];
  }
  if (avoidFerries || avoidHighways || avoidTolls) {
    NSString *buffer = @"";
    int cnt = 0;
    if (avoidFerries) {
      buffer = @"ferries";
      cnt = 1;
    }
    if (avoidTolls) {
      if (cnt > 0) {
        buffer = [buffer stringByAppendingString:@"|"];
      }
      buffer = [buffer stringByAppendingString:@"tolls"];
      cnt++;
    }
    if (avoidHighways) {
      if (cnt > 0) {
        buffer = [buffer stringByAppendingString:@"|"];
      }
      buffer = [buffer stringByAppendingString:@"highways"];
    }
    [params setObject:buffer forKey:@"avoid"];
  }
  
  //-------------------
  // language parameter
  //-------------------
  NSArray<NSString *> *preferredLanguages = [NSLocale preferredLanguages];
  NSString *localeCode = [preferredLanguages objectAtIndex:0];
  NSString *languageCode = [[localeCode componentsSeparatedByString:@"-"] firstObject];
  [params setObject:languageCode forKey:@"language"];

  //-------------------
  // units parameter
  //-------------------
  if ([opts objectForKey:@"unitSystem"]) {
    [params setObject:[[[opts objectForKey:@"unitSystem"] stringValue] lowercaseString] forKey:@"units"];
//  } else if ([@"en_US" isEqualToString:localeCode]) {
//    [params setObject:@"imperial" forKey:@"units"];
//  } else {
//    [params setObject:@"metric" forKey:@"units"];
  }
  
  //-------------------
  // region parameter
  //-------------------
  if ([opts objectForKey:@"region"]) {
    [params setObject:[[[opts objectForKey:@"region"] stringValue] lowercaseString] forKey:@"region"];
  }
  
  [self httpGetWithDictionary:params command:command];
}

- (void)httpGetWithDictionary:(NSDictionary *)params command:(CDVInvokedUrlCommand *)command
{
  
  NSString *urlStr = @"https://maps.googleapis.com/maps/api/directions/json?";
  
  NSUserDefaults *myDefaults = [[NSUserDefaults alloc] initWithSuiteName:@"cordova.plugin.googlemaps"];
  NSString *DEFAULT_APIKey = [myDefaults objectForKey:@"GOOGLE_MAPS_API_KEY"];
  urlStr = [urlStr stringByAppendingFormat:@"key=%@&", DEFAULT_APIKey];

  [self.executeQueue addOperationWithBlock:^{
    [PluginUtil getJsonWithURL:urlStr params:params completionBlock:^(BOOL succeeded, NSDictionary *result, NSString *error) {

      CDVPluginResult* pluginResult;
      if (!succeeded) {
        NSLog(@"[directions] %@", error);
        pluginResult = [CDVPluginResult resultWithStatus: CDVCommandStatus_ERROR messageAsString:error];
      } else {
        pluginResult = [CDVPluginResult resultWithStatus: CDVCommandStatus_OK messageAsDictionary:result];
      }
      [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
  }];
}

@end

File: src/android/plugin/google/maps/PluginDirectionsService.java
package plugin.google.maps;

import android.app.Activity;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.google.android.gms.maps.model.LatLng;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.List;
import java.util.Locale;

public class PluginDirectionsService extends CordovaPlugin {
  private Activity activity;
  private final String TAG = "PluginDirectionsService";
  private String API_KEY = "";

  public void initialize(final CordovaInterface cordova, final CordovaWebView webView) {
    super.initialize(cordova, webView);
    activity = cordova.getActivity();

    try {
      ApplicationInfo ai = activity.getPackageManager().getApplicationInfo(activity.getPackageName(), PackageManager.GET_META_DATA);
      this.API_KEY = ai.metaData.getString("com.google.android.geo.API_KEY");
    } catch (Exception e) {
      // ignore
    }
  }


  @Override
  public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {

    HashMap<String, String> params = new HashMap<String, String>();

    JSONObject opts = args.getJSONObject(0);

    //-------------------
    // required parameters
    //-------------------
    params.put("origin", _decode_DirectionsRequestLocation(opts.getJSONObject("origin")));
    params.put("destination", _decode_DirectionsRequestLocation(opts.getJSONObject("destination")));
    params.put("key", this.API_KEY);

    //-------------------
    // mode parameter
    //-------------------
    if (opts.has("travelMode")) {
      // Default : driving
      String travelMode = opts.getString("travelMode").toLowerCase();
      if (!"driving".equals(travelMode)) {
        params.put("mode", travelMode);

        //-------------------
        // transitOptions parameter
        //-------------------
        if ("transit".equals(travelMode) && opts.has("transitOptions")) {
          JSONObject transitOptions = opts.getJSONObject("transitOptions");
          if (transitOptions.has("arrivalTime")) {
            params.put("arrival_time", String.valueOf(transitOptions.getInt("arrivalTime")));
          }
          if (transitOptions.has("departureTime")) {
            params.put("departure_time", String.valueOf(transitOptions.getInt("departureTime")));
          }
        }
      } else {

        //-------------------
        // DrivingOptions parameter
        //-------------------
        if (opts.has("drivingOptions")) {
          JSONObject drivingOptions = opts.getJSONObject("drivingOptions");
          if (drivingOptions.has("departureTime")) {
            params.put("departure_time", String.valueOf(drivingOptions.getInt("departureTime")));
          }
          if (drivingOptions.has("trafficModel")) {
            String trafficModel = drivingOptions.getString("trafficModel").toLowerCase();
            if (!"best_guess".equals(trafficModel)) {
              params.put("traffic_model", trafficModel);
            }
          }
        }
      }

      //-------------------
      // transitOptions parameter
      //-------------------
      if (opts.has("transitOptions")) {
        JSONObject transitOptions = opts.getJSONObject("transitOptions");
        //-------------------
        // transitOptions.modes parameter
        //-------------------
        if (transitOptions.has("modes")) {
          JSONArray modes = transitOptions.getJSONArray("modes");
          StringBuilder buffer = new StringBuilder();
          int n = modes.length();
          for (int i = 0; i < n; i++) {
            if (i > 0) {
              buffer.append("|");
            }
            buffer.append(modes.getString(i).toLowerCase());
          }
          params.put("transit_mode", buffer.toString());
        }
        //-------------------
        // transitOptions.routingPreference parameter
        //-------------------
        if (transitOptions.has("routingPreference")) {
          String routingPreference = transitOptions.getString("routingPreference").toLowerCase();
          params.put("transit_routing_preference", routingPreference);
        }
      }
    }



    //-------------------
    // waypoints parameter
    //-------------------
    if (opts.has("waypoints")) {
      JSONArray waypoints = opts.getJSONArray("waypoints");
      StringBuilder buffer = new StringBuilder();
      int n = waypoints.length();

      int cnt = 0;
      if (opts.has("optimizeWaypoints") && opts.getBoolean("optimizeWaypoints")) {
        cnt = 1;
        buffer.append("optimize:true");
      }
      for (int i = 0; i < n; i++) {
        JSONObject point = waypoints.getJSONObject(i);
        if (point.has("location")) {
          boolean stopOver = false;
          if (point.has("stopover")) {
            stopOver = point.getBoolean("stopover");
          }
          if (cnt > 0) {
            buffer.append("|");
          }
          if (!stopOver) {
            buffer.append("via:");
          }
          buffer.append(_decode_DirectionsRequestLocation(point.getJSONObject("location")));
          cnt++;
        }
      }
      if (cnt > 0) {
        params.put("waypoints", buffer.toString());
      }
    }
    //-------------------
    // alternatives parameter
    //-------------------
    if (opts.has("provideRouteAlternatives") && opts.getBoolean("provideRouteAlternatives")) {
      params.put("alternatives", "true");
    }

    //-------------------
    // avoid parameter
    //-------------------
    boolean avoidFerries = false;
    boolean avoidHighways = false;
    boolean avoidTolls = false;
    if (opts.has("avoidFerries")) {
      avoidFerries = opts.getBoolean("avoidFerries");
    }
    if (opts.has("avoidHighways")) {
      avoidHighways = opts.getBoolean("avoidHighways");
    }
    if (opts.has("avoidTolls")) {
      avoidTolls = opts.getBoolean("avoidTolls");
    }
    if (avoidFerries || avoidHighways || avoidTolls) {
      StringBuilder buffer = new StringBuilder();
      if (avoidFerries) {
        buffer.append("ferries|");
      }
      if (avoidTolls) {
        buffer.append("tolls|");
      }
      if (avoidHighways) {
        buffer.append("highways|");
      }
      buffer.delete(buffer.length() - 2, buffer.length());
      params.put("avoid", buffer.toString());
    }
    //-------------------
    // language parameter
    //-------------------
    String lCode = Locale.getDefault().getLanguage();
    params.put("language", lCode);

    //-------------------
    // units parameter
    //-------------------
    String localeFull = Locale.getDefault().getDisplayLanguage();
    if (opts.has("unitSystem")) {
      params.put("units", opts.getString("unitSystem").toLowerCase());
    } else if ("en_US".equals(localeFull)) {
      params.put("units", "imperial");
    } else {
      params.put("units", "metric");
    }

    //-------------------
    // region parameter
    //-------------------
    if (opts.has("region")) {
      params.put("region", opts.getString("region").toLowerCase());
    }


    AsyncGetJsonWithURL httpGet = new AsyncGetJsonWithURL("https://maps.googleapis.com/maps/api/directions/json", new AsyncHttpGetInterface() {
      @Override
      public void onPostExecute(JSONObject result) {
        PluginResult pluginResult = null;

        try {
          if (result == null) {
            pluginResult = new PluginResult(PluginResult.Status.ERROR, "UNKNOWN_ERROR");
          } else {
            if ("OK".equals(result.getString("status"))) {
              pluginResult = new PluginResult(PluginResult.Status.OK, result);
            } else {
              pluginResult = new PluginResult(PluginResult.Status.ERROR, result.getString("status"));
            }
          }
        } catch (Exception e){
          pluginResult = new PluginResult(PluginResult.Status.ERROR, "UNKNOWN_ERROR");
        } finally {
          if (pluginResult == null) {
            pluginResult = new PluginResult(PluginResult.Status.ERROR, "UNKNOWN_ERROR");
          }
          callbackContext.sendPluginResult(pluginResult);
        }
      }
    });

    httpGet.execute(params);
    return true;
  }

  private String _decode_DirectionsRequestLocation(JSONObject position) throws JSONException {
    String positionType = position.getString("type");

    if ("string".equals(positionType)) {
      return position.getString("value");
    }

    JSONObject value = position.getJSONObject("value");
    if ("location".equals(positionType)) {
      return String.format("%f,%f", value.getDouble("lat"), value.getDouble("lng"));
    }
    if (value.has("placeId")) {
      return String.format("place_id:%s", value.getString("placeId"));
    }
    if (value.has("location")) {
      JSONObject location = value.getJSONObject("location");
      return String.format("%f,%f", location.getDouble("lat"), location.getDouble("lng"));
    }
    if (value.has("query")) {
      return value.getString("query");
    }
    return "";
  }
}
